### PR TITLE
CPP: Work on Lossy function result cast query

### PR DESCRIPTION
**Pull request by geoffw0**
Added test, qhelp and tags to this query.  Fixed some false positives.  It appears to produce generally good results (despite the comment that was in the `@description`), so I've given it a `medium` precision - which will result on it being run on LGTM but not displayed by default.  If this part turns out to be controversial I can reduce the precision to `low` or delete it.

@jbj - I'd like your opinion on the precision tag.
@Semmle/doc - please review and make suggestions on the new `qhelp`.

### CODE REVIEW

File: cpp/ql/src/Likely Bugs/Conversion/LossyFunctionResultCast.ql [hunk 1]
```diff
@@ -1,18 +1,39 @@
...
+  fName = "nearbyintl" or
+  fName = "rint" or
+  fName = "rintf" or
+  fName = "rintl" or
+  fName = "round" or
+  fName = "roundf" or
+  fName = "roundl" or
+  fName = "trunc" or
+  fName = "truncf" or
+  fName = "truncl"
```
Comment (jbj) [inline cpp/ql/src/Likely Bugs/Conversion/LossyFunctionResultCast.ql]: My main worry is that this list is that this list isn't exhaustive. It doesn't include user-defined wrappers around rounding functions, for example. I also notice that `pow(2, smallPositiveInteger)` is a common FP.

Comment (geoffw0) [reply cpp/ql/src/Likely Bugs/Conversion/LossyFunctionResultCast.ql]: I didn't find any cases of user-defined wrappers 'in the wild', but I've added this functionality.

I've also add `pow` (and variants) as an exception, provided the first number (base) has a compile time value that is a whole number.

**Comment by geoffw0** [comment]
> It's more or less a compiler warning, and we generally leave that type of query to compilers.

That's a fair criticism.

> the results generally looked good. 

That was my observation as well, hence my instinct to make it run but not displayed by default on LGTM.  I'll remove the `@precision` tag if you're not entirely happy with that though.

**Comment by jbj** [comment]
Testing here: https://lgtm.com/query/351600969242589109/

File: cpp/ql/src/Likely Bugs/Conversion/LossyFunctionResultCast.ql (hunk 1 above)
Comment (jbj) [reply cpp/ql/src/Likely Bugs/Conversion/LossyFunctionResultCast.ql]: I ran the query again on a bunch of LGTM projects and found [this FP](https://lgtm.com/projects/g/davidstutz/superpixels-revisited/snapshot/1c646d416b0c0f07f3ef5a5231f2a07b74668fbe/files/lib_vlfeat/vl/mathop.h?sort=name&dir=ASC&mode=heatmap#L177), where `__builtin_ceil` is called. Apparently, GCC has builtin versions of some of these operations. AFAICT from [the compiler docs](https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html), only the ceil and floor functions are available as builtins, but the [`mathop.h` file](https://github.com/vlfeat/vlfeat/blob/master/vl/mathop.h#L163) also uses `__builtin_lroundf` and apparently gets away with it. Perhaps, to be on the safe side, we should just whitelist all these function names with a `__builtin_` prefix as well.

File: cpp/ql/src/Likely Bugs/Conversion/LossyFunctionResultCast.ql [hunk 2]
```diff
@@ -1,18 +1,65 @@
...
+  whitelistPow(fc) or
+  exists(ReturnStmt rs |
+    rs.getEnclosingFunction() = fc.getTarget() and
+    whiteListWrapped(rs.getExpr())
+  ) or
+  exists(ReturnStmt rs, Variable v |
+    rs.getEnclosingFunction() = fc.getTarget() and
+    rs.getExpr().(VariableAccess).getTarget() = v and
+    whiteListWrapped(v.getAnAssignedValue())
+  )
```
Comment (jbj) [inline cpp/ql/src/Likely Bugs/Conversion/LossyFunctionResultCast.ql]: The wrapper detection is failing to detect [`vim_round`](https://github.com/vim/vim/blob/ebc8402063bbd5475ed348469172d4e3e260ae83/src/evalfunc.c#L10086), and this causes a FP on vim. I suggest using `DataFlow::localFlow`.

File: cpp/ql/src/Likely Bugs/Conversion/LossyFunctionResultCast.ql (hunk 1 above)
Comment (geoffw0) [reply cpp/ql/src/Likely Bugs/Conversion/LossyFunctionResultCast.ql]: Thanks for investigating.  I've added an exception for `__builtin_*` as well.

File: cpp/ql/src/Likely Bugs/Conversion/LossyFunctionResultCast.ql (hunk 2 above)
Comment (geoffw0) [reply cpp/ql/src/Likely Bugs/Conversion/LossyFunctionResultCast.ql]: Good thinking, done.

**Pull request merged**